### DISCUSSION TITLE
Unmounting Controller with formState.isValid issues

### DISCUSSION THREAD

**Discussion by Marcin-Kapusta** [react-hook-form/react-hook-form#3643] [Q&A]
I'm using react-hook-form. I'm trying to get rid of some problem that I'm facing at the moment.

The problem is when I need to monitor `formState.isValid` property and when I'm using `Controller` like in example below:

```
import React from 'react';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { useForm, Controller } from 'react-hook-form';

interface ITest { name: string }
const initialData: ITest = { name: 'John' };
const schema: yup.ObjectSchema<ITest> = yup.object<ITest>( {
  name: yup.string().required( 'Name is required' ),
} ).required();

const FormTest: React.FC = () => {
  const [ mount, setMount ] = React.useState<boolean>( true );
  const toggleMount = React.useCallback( () => {
    setMount( ( prevState ) => !prevState );
  }, [] );

  return (
    <div>
      { mount && <SubComponent /> }
      <button type="button" onClick={ toggleMount }>Toggle Form</button>
    </div>
  );
};

export const SubComponent: React.FC = () => {
  const formContext = useForm<ITest>( {
    defaultValues: initialData,
    mode: 'onTouched',
    resolver: yupResolver( schema ),
  } );
  const { handleSubmit, formState, control } = formContext;

  const onSubmit = ( data: ITest ) => {
    console.log( 'submit', data );
  };

  const saveAction = React.useCallback( () => {
    handleSubmit( onSubmit, ( e ) => {
      console.log( 'error', e );
    } )();
  }, [ handleSubmit ] );

  console.log( formState.isValid ); // When remove this line there is no problem during unmounting

  return (
    <form noValidate>
      <Controller name="name"control={ control } defaultValue={ initialData.name } as="input" />
      <button type="button" onClick={ saveAction }>Submit</button>
    </form>
  );
};
```
I provided Button with Label Toggle Form to unmount and mount `SubComponent`. When unmounting I have such error in console:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in SubComponent (at FormalTest.tsx:20)
```
I spent almost whole day trying to figure out what is causing this problem. When I don't use Controller but just pure:

`<input name="name" ref={ register } type="text" defaultValue={ initialData.name } />`
There is no problem at all. But when I start using `Controller` I have this problem during unmounting. I''ve tried to put it very simple but I need to use in my more complex scenario Controller later. Also the problem dissapears when I remove line with monitoring `formState.isValid` but again I need to monitor it and enable/disable some buttons based on this `isValid` flag.

Any advice how I can get rid of this error during unmounting and in the same time use `formState.isValid` and UI components wrapped in `Controller`?

**Comment by bluebill1049** [comment]
can you share a codesandbox with this? here is the reason for the warning: https://react-hook-form.com/faqs#Unmountedcomponentwarning

**Reply by Marcin-Kapusta** [reply]
Can You give me some advice how to create such sandbox? Where to start?

**Reply by bluebill1049** [reply]
here you can use this link: https://codesandbox.io/s/react-hook-form-template-ttzn8

**Reply by Marcin-Kapusta** [reply]
Here You go! https://codesandbox.io/s/reack-hook-form-problem-8upfd
This is my first sandbox. Just press on `Toggle Form` to see error in console. It is happening only during first time after load.

**Reply by bluebill1049** [reply]
Thanks, will get that fixed. It's not a prod issue by the way. I think we need the unmount check in dev build, otherwise, this warning is annoying.

**Reply by bluebill1049** [reply]
will patch in the next release: https://github.com/react-hook-form/react-hook-form/pull/3645

**Reply by Marcin-Kapusta** [reply]
Thanks for understanding and Thank You for fixing it.

**Reply by yhonchisdev** [reply]
Good morning I have the same error, but I get it when I put it with custom steps and when I write in the change I get the same error.

**Comment by yhonchisdev** [comment]
Good morning I have the same error, but I get it when I put it with custom steps and when I write in the change I get the same error.

**Comment by morskibg** [comment]
I have the same error in  realization of multi form via MUI Stepper. The error arises only in first transition from first to second form.
